### ISSUE TITLE
Better diagnostic for collected delegate

### ISSUE THREAD

**Issue by kbaladurin** [dotnet/coreclr#15465]
Current version of CoreCLR contains code for supporting MDAs. It is compiled and some parts of it were removed as dead. MDAs are very useful to detect application bugs that related to transitions beetween managed and unmanaged code, for example, to detect calls of collected delegates.

I've tried to enable current implementation of MDAs in coreclr (https://github.com/dotnet/coreclr/pull/15464):
- I've added `xmlparser` project to coreclr
- I've added some previously deleted files (`mda.cpp`, `mdadac.cpp`, etc.)

After several fixes CallbackOnCollectedDelegate MDA starts work for me:
```
$ COMPlus_MDA=1 ./corerun delegategc.dll
Press ESC to stop
register_callback:0x7fe44088316c ----
force GC!!!
upcall ----
A callback was made on a garbage collected delegate of type 'delegategc!DelegateGC.Program+Callback::Invoke'. This may cause application crashes, corruption and data loss. When passing delegates to unmanaged code, they must be kept alive by the managed application until it is guaranteed that they will never be called.

Unhandled Exception: System.NullReferenceException: Object reference not set to an instance of an object.
   at System.StubHelpers.StubHelpers.CheckCollectedDelegateMDA(IntPtr pEntryThunk)
   at DelegateGC.Program.upcall()
   at DelegateGC.Program.MyView_KeyEvent(ConsoleKeyInfo keyInfo) in /home/kbaladurin/Desktop/dotnet/coredumps/lldb/dotnet/delegategc/Program.cs:line 65
   at DelegateGC.Program.Main(String[] args) in /home/kbaladurin/Desktop/dotnet/coredumps/lldb/dotnet/delegategc/Program.cs:line 85
Aborted (core dumped)
```

Is it correct approach to enable MDAs or have you another plans about it?

Thank you!

**Comment by kbaladurin** [comment]
cc @Dmitri-Botcharnikov @ayuckhulk

**Comment by kbaladurin** [comment]
@janvorli @jkotas PTAL

**Comment by jkotas** [comment]
What are the MDA you are really interested in? Is the CollectedDelegateMDA the only one?

**Comment by jkotas** [comment]
cc @noahfalk @lt72

**Comment by kbaladurin** [comment]
@jkotas yes, now I'm interested in CollectedDelegateMDA because we often face this problem. But there are other assistants that can be helpful, I think.

**Comment by jkotas** [comment]
This diagnostic for calls on collected delegates should be on by default.  There is no good reason why it needs to be a special runtime knob. @parjong started working towards it in #12731 by introducing `UMEntryThunkCode::Poison`. It will detect the calls on collected delegate, but there is no nice error message. It should be pretty straightfoward to make it better by:
- Changing it to produce better error message
- Changing the allocator for UM thunks from LIFO to FIFO so that it takes longer before the thunks get reused.
Doing these two things will make it better than what you get out of the MDA.

**Comment by lt72** [comment]
@davmason @sywhang: let's pick this up for 2.1

**Comment by kbaladurin** [comment]
@jkotas thank you for suggestion! I'll try to use this approach.

**Comment by jkotas** [comment]
Fixed by #15809
